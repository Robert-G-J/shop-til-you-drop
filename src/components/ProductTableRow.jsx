import React from "react";
import PropTypes from "prop-types";

const ProductTableRow = ({
  id,
  name,
  description,
  price,
  image,
  quantity,
  updateCart
}) => (
  <tr>
    <td className="image bp-tablet">
      <img src={image} alt={name} />
    </td>

    <td className="product">
      {name}
      <p className="description bp-tablet">{description}</p>
    </td>
    <td className="price ">£{price}</td>
    <td className="quantity">
      <input
        required
        className="quantity"
        type="number"
        name="quantity"
        min="0"
        value={quantity}
        onChange={e => updateCart(id, parseInt(e.target.value, 10))}
      />
    </td>
    <td className="total">£{(price * quantity).toFixed(2)}</td>
  </tr>
);

ProductTableRow.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  quantity: PropTypes.number.isRequired,
  updateCart: PropTypes.func.isRequired
};

ProductTableRow.defaultProps = {
  quantity: 0
};
export default ProductTableRow;
