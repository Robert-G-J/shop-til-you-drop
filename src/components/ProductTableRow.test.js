import React from "react";
import ProductTableRow from "./ProductTableRow";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";

describe("<ProductTableRow />", () => {
  let component;
  let mockChange;
  const props = {
    id: 1,
    name: "bumblescnark",
    price: "2.10",
    description:
      "A scnarfing good time throughout any passive aggressive encounter",
    quantity: 1,
    image: "https://i.ytimg.com/vi/2GlCcO4zghM/hqdefault.jpg"
  };

  beforeEach(() => {
    mockChange = jest.fn();
    component = shallow(
      <ProductTableRow {...props} updateQuantity={mockChange} />
    );
  });

  it("renders correctly", () => {
    const tree = renderer
      .create(<ProductTableRow {...props} updateQuantity={mockChange} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("doesn't crash", () => {
    expect(component.exists()).toEqual(true);
  });

  it("has product id", () => {
    expect(component.find("td.id").length).toEqual(1);
  });

  it("has product name", () => {
    expect(component.find("td.product").length).toEqual(1);
  });

  it("has product description", () => {
    expect(component.find("td .description").length).toEqual(1);
  });

  it("has product price", () => {
    expect(component.find("td.price").length).toEqual(1);
  });

  it("has a product image", () => {
    expect(component.find("td.image img").length).toEqual(1);
  });
  it("has product quantity cell", () => {
    expect(component.find("td.quantity").length).toEqual(1);
  });

  describe("<input>", () => {
    it("exists", () => {
      expect(component.find("input.quantity").length).toEqual(1);
    });
    xit("should call a click handler onChange", () => {
      component.find(".input-quantity").simulate("change");
      expect(mockChange).toHaveBeenCalled();
      // TODO: fix test- event is undefined in <input/> onChange. Mounting makes enzyme wrap the <tr/> in a div, also raising an error
    });
  });
});
