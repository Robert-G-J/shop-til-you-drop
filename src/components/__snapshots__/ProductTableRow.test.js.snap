// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTableRow /> renders correctly 1`] = `
<tr>
  <td
    className="id"
  >
    1
  </td>
  <td
    className="image"
  >
    <img
      alt="bumblescnark"
      src="https://i.ytimg.com/vi/2GlCcO4zghM/hqdefault.jpg"
    />
  </td>
  <td
    className="product"
  >
    bumblescnark
    <p
      className="description"
    >
      A scnarfing good time throughout any passive aggressive encounter
    </p>
  </td>
  <td
    className="price "
  >
    £
    2.10
  </td>
  <td
    className="quantity"
  >
    <input
      className="quantity"
      min="0"
      name="quantity"
      onChange={[Function]}
      type="number"
      value={1}
    />
  </td>
  <td
    className="total"
  >
    £
    2.10
  </td>
</tr>
`;
